### PR TITLE
[SLP] Normalize debug messages for newTreeEntry.

### PR DESCRIPTION
**Pull request by HanKuanChen**
A debug message should follow after newTreeEntry.
Make ExtractValueInst and ExtractElementInst use setOperand directly.

### CODE REVIEW

**Review by alexey-bataev — COMMENTED** [review]
Would be useful to dump the list of scalars, if possible, helps in many cases

**Comment by alexey-bataev** [comment]
> > Would be useful to dump the list of scalars, if possible, helps in many cases
> 
> how about dump `TreeEntry` directly?

Sounds good

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -3671,6 +3671,7 @@ class BoUpSLP {
 
     if (UserTreeIdx.UserTE)
       Last->UserTreeIndices.push_back(UserTreeIdx);
+    Last->dump();
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3674]: No, it's a bad solution. 1. It first be emitted only in debug builds and only if -debug is provided. 2. Better to move it to each LLVM_DEBUG, where the messages for entries are printed

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -3671,7 +3671,7 @@ class BoUpSLP {
 
     if (UserTreeIdx.UserTE)
       Last->UserTreeIndices.push_back(UserTreeIdx);
-    Last->dump();
+    LLVM_DEBUG(Last->dump());
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3674]: No, make it part of debug messages, do not dump it here

**Review by alexey-bataev — APPROVED** [review]

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -8417,13 +8418,14 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
...
+                                   ReuseShuffleIndices, CurrentOrder);
+      LLVM_DEBUG(dbgs() << "SLP: added a new TreeEntry "
+                           "(ExtractValueInst/ExtractElementInst).\n";
+                 TE->dump());
       // This is a special case, as it does not gather, but at the same time
       // we are not extending buildTree_rec() towards the operands.
-      ValueList Op0;
-      Op0.assign(VL.size(), VL0->getOperand(0));
-      VectorizableTree.back()->setOperand(0, Op0);
+      TE->setOperand(*this);
```
Comment (alexfh) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: Was this change intended for this commit? It's not just about debug messages, IIUC.

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: We see test failures in code compiled by clang built after this commit. I wonder if this is related to the this snippet. While I'm working on a test case I briefly looked at this commit and found only one suspicious fragment - this one.

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: Can you send me the test?

Comment (asmok-g) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: In another instance, I think this is causing an infinite loop. Working on some repro

Comment (asmok-g) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: I don't claim to have any knowledge of the code here; but did it accidentally remove the handling of a special case, according to the comment in the code ?

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: This PR should not cause infinite loop. I need the failure test.

Comment (asmok-g) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: My bad, I double checked and made sure the infinite loop is reproducible without the patch.

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: Here a reduced test that shows the difference in produced IR after this commit. I'm not sure if it's correct or not, but I believe this was not intended. The patch might have been over-reduced (since I used the presence of any diff as the interestingness test), but the original code seems to be correct, but after this commit it starts producing a segfault.

```
$ cat reduced.ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define void @_f() #0 {
  %1 = extractelement <8 x float> zeroinitializer, i64 1
  %2 = fcmp uno float %1, 0.000000e+00
  %3 = select i1 %2, i64 0, i64 0
  %4 = extractelement <8 x float> zeroinitializer, i64 2
  %5 = fcmp uno float %4, 0.000000e+00
  %6 = select i1 %5, i64 0, i64 0
  %7 = extractelement <8 x float> zeroinitializer, i64 3
  %8 = fcmp uno float %7, 0.000000e+00
  %9 = extractelement <8 x float> zeroinitializer, i64 4
  %10 = fcmp uno float %9, 0.000000e+00
  %11 = select i1 %10, i64 0, i64 0
  %12 = extractelement <8 x float> zeroinitializer, i64 5
  %13 = fcmp uno float %12, 0.000000e+00
  %14 = extractelement <8 x float> zeroinitializer, i64 6
  %15 = fcmp uno float %14, 0.000000e+00
  %16 = select i1 %15, i64 0, i64 0
  %17 = extractelement <8 x float> zeroinitializer, i64 7
  %18 = fcmp uno float %17, 0.000000e+00
  %19 = fcmp oeq float %12, 0.000000e+00
  %20 = fcmp oeq float %14, 0.000000e+00
  %21 = select i1 %20, i64 0, i64 0
  %22 = fcmp oeq float %17, 0.000000e+00
  %23 = select i1 %22, i64 0, i64 0
  %24 = fcmp oeq float %4, 0.000000e+00
  %25 = select i1 %24, i64 0, i64 0
  %26 = fcmp oeq float %7, 0.000000e+00
  %27 = select i1 %26, i64 0, i64 0
  %28 = fcmp oeq float %9, 0.000000e+00
  %29 = fcmp oeq float %12, 0.000000e+00
  %30 = select i1 %29, i64 0, i64 0
  %31 = fcmp oeq float %14, 0.000000e+00
  %32 = select i1 %31, i64 0, i64 0
  %33 = fcmp oeq float %17, 0.000000e+00
  ret void
}

attributes #0 = { "target-features"="+aes,+avx,+avx2,+avx512bw,+avx512dq,+avx512f,+avx512vl,+bmi2,+cmov,+crc32,+cx16,+cx8,+evex512,+f16c,+fma,+fxsr,+lzcnt,+mmx,+pclmul,+popcnt,+prfchw,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87,+xsave" }
$ diff -u <(./opt-before --passes=slp-vectorizer reduced.ll -S -o -) <(./opt-after --passes=slp-vectorizer reduced.ll -S -o -)
--- /dev/fd/63  2025-01-14 17:46:34.291801310 +0100
+++ /dev/fd/62  2025-01-14 17:46:34.291801310 +0100
@@ -4,16 +4,16 @@
 target triple = "x86_64-unknown-linux-gnu"

 define void @_f() #0 {
-  %1 = select i1 false, i64 0, i64 0
-  %2 = select i1 false, i64 0, i64 0
-  %3 = select i1 false, i64 0, i64 0
-  %4 = select i1 false, i64 0, i64 0
-  %5 = select i1 true, i64 0, i64 0
-  %6 = select i1 true, i64 0, i64 0
-  %7 = select i1 true, i64 0, i64 0
-  %8 = select i1 true, i64 0, i64 0
-  %9 = select i1 true, i64 0, i64 0
-  %10 = select i1 true, i64 0, i64 0
+  %1 = select i1 poison, i64 0, i64 0
+  %2 = select i1 poison, i64 0, i64 0
+  %3 = select i1 poison, i64 0, i64 0
+  %4 = select i1 poison, i64 0, i64 0
+  %5 = select i1 poison, i64 0, i64 0
+  %6 = select i1 poison, i64 0, i64 0
+  %7 = select i1 poison, i64 0, i64 0
+  %8 = select i1 poison, i64 0, i64 0
+  %9 = select i1 poison, i64 0, i64 0
+  %10 = select i1 poison, i64 0, i64 0
   ret void
 }
```

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: I also confirmed that this particular snippet is responsible for the change in the slp-vectorizer output. I.e. undoing it fixes the problem:
```
8449c8449,8452
<       TE->setOperand(*this);
---
>       (void)TE;
>       ValueList Op0;
>       Op0.assign(VL.size(), VL0->getOperand(0));
>       VectorizableTree.back()->setOperand(0, Op0);
```

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: I can reliably reproduce this and here's the same example in compiler explorer: https://gcc.godbolt.org/z/cEWPchGaE

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: I mean, I could reliably reproduce this with opt built right before and at the same commit. I haven't tried it with https://github.com/llvm/llvm-project/commit/539b15b41a6a01017c0a555e89b7d2b62ba194d2

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: I confirmed this at b270525f730be6e7196667925f5a9bfa153262e9, but at 283dca56f8dddbf2f144730a01675c94b04f57cb this already seems to have been fixed. Searching for the fix.

Comment (alexfh) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8428]: Apparently, 066b88879ab5c195e7e14609e546cc238c2f3bf3 fixes this.
